react-ui/src/pages/contest/ContestAppBar.jsx

```diff
@@ -0,0 +1,112 @@
+/**
+ * Contest-specific content for top app bar
+ */
+
+import Box from '@material-ui/core/Box';
+import EmojiEventsOutlinedIcon from '@material-ui/icons/EmojiEventsOutlined';
+import SettingsOutlinedIcon from '@material-ui/icons/SettingsOutlined';
+import ThumbsUpDownOutlinedIcon from '@material-ui/icons/ThumbsUpDownOutlined';
+import PropTypes from 'prop-types';
+import { useLocation, useParams } from 'react-router-dom';
+
+import {
+  useSwrData,
+} from '../../common';
+import {
+  ArrowBackButton,
+  CustomIconButton,
+  Countdown,
+} from '../../components';
+
+function ContestAppBarMain({ handleVotingExpired, handleReload }) {
+  const { contestId } = useParams();
+  const { state = {} } = useLocation();
+
+  const apiPath = `/contests/${contestId}`;
+  const { data: contest } = useSwrData(apiPath, false);
+
+  const backLink = (state || {}).back || '/contests';
+
+  const {
+    date,
+    isContestMode,
+    voteEnd,
+  } = contest;
+  const voteEndDate = new Date(voteEnd);
+
+  return (
+    <>
+      <ArrowBackButton state={{ date }} to={backLink} />
+      {isContestMode && (
+        <Box display="inline-flex" paddingLeft={1.5}>
+          <Countdown
+            endDate={voteEndDate}
+            handleExpiry={handleVotingExpired}
+            handleReload={handleReload}
+          />
+        </Box>
+      )}
+    </>
+  );
+}
+
+function ContestAppBarRight({ setInfoOpen }) {
+  const { contestId } = useParams();
+
+  const apiPath = `/contests/${contestId}`;
+  const { data: contest } = useSwrData(apiPath, false);
+
+  const toggleSettingsOpen = () => {
+    setInfoOpen((prev) => !prev);
+  };
+
+  const {
+    localVoting,
+    validRedditId,
+    winnersThreadId,
+  } = contest;
+
+  return (
+    <>
+      {!localVoting && validRedditId && (
+        <CustomIconButton
+          href={`https://redd.it/${contestId}`}
+          ariaLabel="Open voting thread"
+          Icon={ThumbsUpDownOutlinedIcon}
+        />
+      )}
+      {winnersThreadId && (
+        <CustomIconButton
+          href={`https://redd.it/${winnersThreadId}`}
+          ariaLabel="Open winners thread"
+          Icon={EmojiEventsOutlinedIcon}
+        />
+      )}
+      <CustomIconButton
+        ariaLabel="View settings"
+        onClick={toggleSettingsOpen}
+        Icon={SettingsOutlinedIcon}
+      />
+    </>
+  );
+}
+
+ContestAppBarRight.propTypes = {
```

Comment:
If we're going to have multiple components in the same file, could we keep the order of the PropTypes definitions in the same order of the components?

I guess one of the reasons I'm not the biggest fan of multiple components in the same file is it's a bit of a cognitive load when you're trying to add/change a prop in `ContestAppBarMain` and then have to scroll down here and figure out which of the two similar looking component names `.propTypes` needs to be edited. If it's a single file, you just look for `propTypes` and edit.

Maybe an argument to start phasing in TypeScript and we won't need `PropTypes` at all? I think TS/JS, TSX/JSX can coexist in the same app with the right configuration.

Reply:
TypeScript might be ideal, I'm also positive you can just upgrade individual files to ts over time without issue.

But ya, I'll separate AppBarMain and AppBarRight... it's just weird because hierarchically I want there to be Contest > AppBar > AppBarMain && AppBarRight. But consistency dictates they be separate files.

Reply:
_(Optional)_ You could consider adding a folder for the hierarchy, but that may be more effort than it's worth.